### PR TITLE
Shader generator improvements

### PR DESCRIPTION
**Pull request by guusw**
Various work on shader translator.

Composition now happens whenever the renderer requests a shader variant, this allows composition based on actual parameter, input/output types and available fields.
It also removes the need to explicitly specify the expected type and dimension on some shader shards.

This PR also adds a bunch of shader features (If, ForEach, Comparisons) and some fixes for exisiting shader shards.

NOTE: some shards have been moved to header files with public fields so they can be read by the shader translator (flow.hpp, wires.hpp)

### CODE REVIEW

**Comment by guusw** [comment]
Adding some points since it involves some additional work besides just the issue that was linked

File: src/core/shards/wires.cpp [hunk 1]
```diff
@@ -11,254 +10,197 @@
 #endif
 
 namespace shards {
-enum RunWireMode { Inline, Detached, Stepped };
 
-struct WireBase {
```
Comment (guusw) [inline src/core/shards/wires.cpp:16]: Moved into header

File: src/core/shards/flow.cpp [hunk 2]
```diff
@@ -1,1134 +1,9 @@
...
 /* Copyright © 2019 Fragcolor Pte. Ltd. */
 
-#include "shardwrapper.hpp"
-#include "shards.h"
-#include "foundation.hpp"
-#include "shared.hpp"
+#include "flow.hpp"
 
 namespace shards {
-static Type condShardSeqs = Type::SeqOf(CoreInfo::ShardsOrNone);
```
Comment (guusw) [inline src/core/shards/flow.cpp:10]: Moved into header

**Comment by guusw** [comment]
Have to squash some WIP commits but otherwise ready

**Review by sinkingsugar — APPROVED** [review]
I think we need a lot of tests,
I could not see if there were major changes in flow or wires, I hope not.

**Comment by guusw** [comment]
@sinkingsugar Not changes in flow/wire shards besides making some fields public.

**Comment by Kryptos-FR** [comment]
Why is it not targeting `devel`?

**Comment by guusw** [comment]
I will make a new branch with doc additions (shard help, examples, small explanations, etc.)

**Pull request merged**